### PR TITLE
Implement preprocessing

### PR DESCRIPTION
**Pull request by cschaefer26**
- added basic preprocessing (tokenization, mel specs)
- scratched training (to be properly done later)

### CODE REVIEW

File: dfa/dataset.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    def __init__(self, item_ids: List[str], mel_dir: Path, token_dir: Path):
+        super().__init__()
+        self.item_ids = item_ids
+        self.mel_dir = mel_dir
+        self.token_dir = token_dir
+
+    def __getitem__(self, index):
+        item_id = self.item_ids[index]
+        mel = torch.load(self.mel_dir / f'{item_id}.pt')
+        tokens = torch.load(self.token_dir / f'{item_id}.pt')
```
Comment (cfrancesco) [inline dfa/dataset.py:26]: numpy objects here might be easier to inspect than torch objects?

File: dfa/model.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            in_channels, out_channels, kernel_size,
+            stride=1, padding=kernel_size // 2, bias=False)
+        self.bnorm = nn.BatchNorm1d(out_channels)
+        self.relu = nn.ReLU()
+
+    def forward(self, x):
+        x = x.transpose(1, 2)
+        x = self.conv(x)
+        x = self.relu(x)
+        x = self.bnorm(x)
```
Comment (cfrancesco) [inline dfa/model.py:19]: bnorm should be before relu

File: dfa/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    def __call__(self, wav_path: Path) -> Dict[str, Union[str, int]]:
+        item_id = wav_path.stem
+        wav = self.audio.load_wav(wav_path)
+        mel = self.audio.wav_to_mel(wav)
+        mel = torch.tensor(mel).float()
+        text = self.text_dict[item_id]
+        tokens = self.tokenizer(text)
+        tokens = torch.tensor(tokens).long()
+        torch.save(mel, self.paths.mel_dir / f'{item_id}.pt')
+        torch.save(tokens, self.paths.token_dir / f'{item_id}.pt')
```
Comment (cfrancesco) [inline dfa/preprocess.py:35]: Root of the comment above, numpy object would be  easier to inspect from outside, unless torch objects are more efficient in some way

File: dfa/scratch_pred.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    config = checkpoint['config']
+    symbols = checkpoint['symbols']
+    audio = Audio(**config['audio'])
+    tokenizer = Tokenizer(symbols)
+    model = Aligner.from_checkpoint(checkpoint).eval()
+    print(f'model step {model.get_step()}')
+
+    main_dir = Path('/Users/cschaefe/datasets/audio_data/Cutted_merged')
+    file_id = '04902'
+
```
Comment (cfrancesco) [inline dfa/scratch_pred.py]: would be nice to already use the Paths object

File: dfa/scratch_pred.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    wav = audio.load_wav(main_dir / f'{file_id}.wav')
+    mel = audio.wav_to_mel(wav)
+    mel = torch.tensor(mel).float().unsqueeze(0)
+
+    pred = model(mel)
+
+    pred[:, :, 0] = -9999 # remove pad pred
```
Comment (cfrancesco) [inline dfa/scratch_pred.py]: what's this? :D 
`pred = pred[:,:,1:]`?

File: dfa/scratch_train.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    ctc_loss = CTCLoss()
+    tokenizer = Tokenizer(symbols)
+    dataloader = new_dataloader(dataset_path=paths.data_dir / 'dataset.pkl', mel_dir=paths.mel_dir,
+                                token_dir=paths.token_dir, batch_size=16)
+
+    for epoch in range(1, 1000):
+        for i, batch in enumerate(dataloader):
+            tokens, mel, tokens_len, mel_len = to_device(batch, device)
+            pred = model(mel)
+            pred = pred.transpose(0, 1).log_softmax(2)
```
Comment (cfrancesco) [inline dfa/scratch_train.py:58]: move this line into model?

**Review by cfrancesco — COMMENTED** [review]
I noticed that there is no phoemizer. Is this an explicit choice or just a temp thing?

File: dfa/dataset.py (hunk 1 above)
Comment (cschaefer26) [reply dfa/dataset.py:26]: though they would have been converted to torch tensors every batch

File: dfa/model.py (hunk 2 above)
Comment (cschaefer26) [reply dfa/model.py:19]: Changed

File: dfa/preprocess.py (hunk 3 above)
Comment (cschaefer26) [reply dfa/preprocess.py:35]: Yeah, some overhead with loading torch from numpy

File: dfa/scratch_pred.py (hunk 4 above)
Comment (cschaefer26) [reply dfa/scratch_pred.py]: Absolutely,  thats just the sketch

File: dfa/scratch_pred.py (hunk 5 above)
Comment (cschaefer26) [reply dfa/scratch_pred.py]: Here to remove the pad prediction to see whats the top predicted char

File: dfa/scratch_train.py (hunk 6 above)
Comment (cschaefer26) [reply dfa/scratch_train.py:58]: could be done with a switch for eval mode

**Review by cfrancesco — APPROVED** [review]
